src/data/transform.ts

```diff
@@ -435,124 +427,159 @@ export const convertToExcalidrawElements = (
   if (!elements) {
     return [];
   }
+
   const elementStore = new ElementStore();
-  // Push all elements to array as there could be cases where element id
-  // is referenced before element is created
-  elements.forEach((element) => {
-    let elementId = element.id || regenerateId(null);
+
+  // ensure unique ids
+  // ---------------------------------------------------------------------------
+  const ids = new Set<ExcalidrawElement["id"]>();
+  const elementsWithIds = elements.map((element) => {
+    let id = element.id || regenerateId(null);
 
     // To make sure every element has a unique id since regenerateId appends
     // _copy to the original id and if it exists we need to generate again
     // hence a loop
-    while (elementStore.hasElementWithId(elementId)) {
-      elementId = regenerateId(elementId);
+    while (ids.has(id)) {
+      id = regenerateId(id);
     }
-    const elementWithId = { ...element, id: elementId };
-    elementStore.add(elementWithId as ExcalidrawElement);
+
+    ids.add(id);
+
+    return { ...element, id };
   });
+  // ---------------------------------------------------------------------------
 
-  const pushedElements =
-    elementStore.getElements() as readonly ExcalidrawProgrammaticElement[];
```

Comment:
I've split the id checking/generation into a separate step & array as we shouldn't be pushing a type to an array that doesn't support it

Reply:
All elements are eventually added and this is just an element store so not sure what types you are referring to here but this is currently breaking the code as mentioned above

Reply:
you were pushing programmatic elements into regular elements array which breaks the type contract (hence why you had to assert both here and when looping through the array below). This is not safe and at that point we could as well have an `any[]` array.

Reply:
yes, but the point of TS is so that we (in the present and other people in the future) don't have to put the whole program into their mind to be able to know what's happening and ensure correctness. Making fundamental type contract violations like here is what we shouldn't be doing unnecessarily.